src/main/java/com/uid2/admin/store/writer/EncryptedScopedStoreWriter.java

```diff
@@ -0,0 +1,113 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Map;
+
+public class EncryptedScopedStoreWriter extends ScopedStoreWriter {
+
+    private final RotatingS3KeyProvider s3KeyProvider;
+    private Integer siteId;
+
+    public EncryptedScopedStoreWriter(IMetadataVersionedStore provider,
+                                      FileManager fileManager, VersionGenerator versionGenerator, Clock clock,
+                                      StoreScope scope, FileName dataFile, String dataType, RotatingS3KeyProvider s3KeyProvider, Integer siteId) {
+        super(provider, fileManager, versionGenerator, clock, scope, dataFile, dataType);
+        this.s3KeyProvider = s3KeyProvider;
+        //site id is passed in to look up S3 key to encrypt
+        this.siteId = siteId;
+    }
+
+
+    public void uploadWithEncryptionKey(String data, JsonObject extraMeta, S3Key encryptionKey) throws Exception {
+        byte[] secret = Base64.getDecoder().decode(encryptionKey.getSecret());
+        byte[] encryptedPayload = AesGcm.encrypt(data.getBytes(StandardCharsets.UTF_8), secret);
+        JsonObject encryptedJson = new JsonObject()
+                .put("key_id", encryptionKey.getId())
+                .put("encryption_version", "1.0")
+                .put("encrypted_payload", Base64.getEncoder().encodeToString(encryptedPayload));
+
+        this.upload(encryptedJson.encodePrettily(), extraMeta);
+    }
+
+    @Override
+    public void upload(String data, JsonObject extraMeta) throws Exception {
+        if (isEncrypted(data)) {
+            super.upload(data, extraMeta);
+        } else {
+            if (siteId == null) {
+                throw new IllegalStateException("Site ID is not set.");
+            }
+
+            Map<Integer, S3Key> s3Keys = s3KeyProvider.getAll();
+            S3Key largestKey = null;
+
+            for (S3Key key : s3Keys.values()) {
+                if (key.getSiteId() == siteId) {
+                    if (largestKey == null || key.getId() > largestKey.getId()) {
+                        largestKey = key;
+                    }
+                }
+            }
+
+            if (largestKey != null) {
+                uploadWithEncryptionKey(data, extraMeta, largestKey);
+            } else {
+                throw new IllegalStateException("No S3 keys available for encryption for site ID: " + siteId);
+            }
+        }
+    }
+
+    private boolean isEncrypted(String data) {
+        try {
+            JsonObject json = new JsonObject(data);
+            return json.containsKey("key_id") &&
+                    json.containsKey("encryption_version") &&
+                    json.containsKey("encrypted_payload");
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+
+    public String getDecryptedContent(String encryptedContent) throws Exception {
```

Comment:
Why does the writer need the ability to decrypt?

Reply:
I have this decryption function here for testing

Reply:
will move decryption logics to shared